endpoint: add desired and realized EndpointPolicy fields

ianvernon opened the pull request:
Remove the fields which are now contained within the `EndpointPolicy` type that were previously in Endpoint, and plumb them appropriately throughout the regeneration process for an Endpoint.

Signed-off by: Ian Vernon <ian@cilium.io>

<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/6375)
<!-- Reviewable:end -->

Commit:
endpoint: add {desired,realized}Policy fields

These fields represent the state which has not been plumbed to the datapath,
and the state which has been plumbed to the datapath related to policy.

Remove the L3Policy, L4Policy, and fields representing ingress and egress
policy enforcement from the Endpoint structure now that the desired / realized
policy contain all of these fields grouped together.

Signed-off by: Ian Vernon <ian@cilium.io>

joestringer commented on pkg/endpoint/bpf.go:
```diff
@@ -204,10 +204,10 @@ func (e *Endpoint) writeHeaderfile(prefix string, owner Owner) error {
 	fw.WriteString(e.Options.GetFmtList())
 
-	if e.L3Policy == nil {
+	if (e.desiredPolicy == nil) || (e.desiredPolicy != nil && e.desiredPolicy.CIDRPolicy == nil) {
```
does the latter part here need to check that `e.desiredPolicy != nil`? (I guess this is really a question for golang conditional precedence ordering guarantees)

ianvernon replied:
It does not :-) fixed in follow-up commit!

joestringer commented on pkg/endpoint/bpf.go:
```diff
@@ -334,15 +334,15 @@ func (e *Endpoint) addNewRedirectsFromMap(owner Owner, m policy.L4PolicyMap, des
 				direction = trafficdirection.Egress
 			}
-			keysFromFilter := e.convertL4FilterToPolicyMapKeys(&l4, direction)
+			keysFromFilter := l4.ToKeys(&l4, direction, *e.prevIdentityCache)
```
Not technically part of this PR, but I'm not sure why `ToKeys()` both takes `l4` as a pointer receiver, and passes a pointer to itself in to the function.

ianvernon replied:
Fixed that in a follow-up commit.

joestringer commented on pkg/endpoint/endpoint.go:
```diff
@@ -341,42 +313,58 @@ func (e *Endpoint) HasBPFProgram() bool {
-// SetEgressPolicyEnabled sets Endpoint's egress policy enforcement
+// SetRealizedIngressPolicyEnabled sets Endpoint's ingress policy enforcement
+// configuration to the specified value. The endpoint's mutex must not be held.
+func (e *Endpoint) SetRealizedIngressPolicyEnabled(ingress bool) {
```
Are these `xxxRealizedxxx()` functions supposed to be used somewhere?

I was also looking at this from a separation of interests perspective, it seems like these functions are reaching into the policy implementation details. It'd be neat if for instance the `SetxxxPolicyEnabled()` functions could be shifted down to the `EndpointPolicy()` object, then maybe just have `func (e *Endpoint) SetxxxPolicyEnabled()` which always sets the desired policy (because external callers into the endpoint should never be allowed to set the realized state, that should only be handled by `updateRealizedState()` inside the BPF package)

ianvernon replied:
I'll remove the functions in a follow-up commit as part of this PR.

Your second point is spot-on. Definitely something that will come soon! :D

ianvernon replied:
https://github.com/cilium/cilium/issues/6419

joestringer commented on pkg/endpoint/endpoint.go:
```diff
@@ -707,27 +696,51 @@ func (e *Endpoint) GetPolicyModel() *models.EndpointPolicyStatus {
 		AllowedEgressIdentities:  desiredEgressIdentities,
-		CidrPolicy:               e.L3Policy.GetModel(),
-		L4:                       e.DesiredL4Policy.GetModel(),
+		CidrPolicy:               desiredCIDRPolicy.GetModel(),
```
I realize these things are all incremental steps in the right direction, but would be nice to see some of the dereferencing of fields inside the policy objects to be shifted into the policy package as well, eliminating the boilerplate in the chunk above these local `desiredxxxPolicy` references. I'm imagining these calls could boil down to something like `e.desiredPolicy.GetCIDRModel()` / `e.desiredPolicy.GetL4Model()`.

ianvernon replied:
https://github.com/cilium/cilium/issues/6419

joestringer commented on pkg/endpoint/policy.go:
```diff
@@ -53,45 +52,6 @@ func (e *Endpoint) ProxyID(l4 *policy.L4Filter) string {
-// convertL4FilterToPolicyMapKeys converts filter into a list of PolicyKeys
-// that apply to this endpoint.
-// Must be called with endpoint.Mutex locked.
-func (e *Endpoint) convertL4FilterToPolicyMapKeys(filter *policy.L4Filter, direction trafficdirection.TrafficDirection) []policy.Key {
```
Please keep the newly introduced equivalent in the same PR as the deletion of code in future. It's much easier to review something like this if we can compare the old and new versions side-by-side (ideally in the same commit), which would also help trace the git history of where the new implementation was introduced, and where it came from originally.

joestringer commented on pkg/endpoint/policy.go:
```diff
@@ -435,7 +397,16 @@ func (e *Endpoint) updateRealizedState(stats *regenerationStatistics, origDir st
+	e.realizedPolicy.IngressPolicyEnabled = e.desiredPolicy.IngressPolicyEnabled
+	e.realizedPolicy.EgressPolicyEnabled = e.desiredPolicy.EgressPolicyEnabled
+	e.realizedPolicy.L4Policy = e.desiredPolicy.L4Policy
+	e.realizedPolicy.CIDRPolicy = e.desiredPolicy.CIDRPolicy
```
Why is it necessary to reference individual pieces of the realized policy, rather than for example doing something like the following?
```
e.realizedPolicy = e.desiredPolicy
```
or 
```
e.realizedPolicy = e.desiredPolicy.Copy()
```

ianvernon replied:
The reason I did this is quite subtle (although it mirrors existing behavior), in that I do not set the references to be the same ( `e.realizedPolicy = e.desiredPolicy` ) because that would update the `MapState` fields in both objects to point to the same map. So, in `syncPolicyMap`, updating of the `desiredPolicy` `MapState` would also update the `realizedPolicy` `mapState`. The other fields in `desiredPolicy` are never updated except for during policy generation. 

We could `DeepCopy` all fields to avoid any potential corruption between fields, but I do wonder how that would impact performance, especially in the case of `L4Policy` with a large number of rules, which causes a lot of performance degradation due to the large number of `EndpointSelector` objects in `L4Filter.Endpoints`. 

What do you think?

joestringer replied:
I don't see the point in making a deepcopy if we could ensure that, for example, the `L4Policy` object is not subsequently modified in the type of way you describe above.

I think all I'm really looking for here is to hide this `EndpointPolicy` field knowledge from the endpoint package. We could probably achieve this just by putting these lines behind a call that looks something like `e.realizedPolicy.Realizes(e.desiredPolicy)`:
```
// Realizes provides a way for the caller to update the received policy
// to reflect that it now has the same contents as the provided desired policy.
func (p *EndpointPolicy) Realizes(desired *EndpointPolicy) {
...
}
```

ianvernon replied:
@joestringer good idea - I'll add this to the follow-up ticket!

joestringer reviewed (changes requested):
This seems like a positive incremental step in the direction of cleaving policy-specific bits from endpoint, I think most of my comments below are actually about potential follow-ups.

There's one piece where it seems that completely unused functions are being introduced. I looked also at #6404 and didn't see a user there, so I'm not sure why this PR introduces them.

Commit:
policy: remove L4Filter parameter from ToKeys function

The function has an L4Filter receiver, which makes this parameter useless, and
the function prone to misuse. Remove the paramter accordingly.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
endpoint: remove redundant non-nil check

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
endpoint: remove unused functions

Signed-off by: Ian Vernon <ian@cilium.io>

aanm commented on pkg/endpoint/endpoint.go:
```diff
@@ -707,27 +680,51 @@ func (e *Endpoint) GetPolicyModel() *models.EndpointPolicyStatus {
 	e.proxyStatisticsMutex.RUnlock()
 	sortProxyStats(proxyStats)
 
+	var (
```
You don't need to to this, the `GetModel()` won't panic if the receiver is nil.

aanm commented on pkg/endpoint/policy.go:
```diff
@@ -160,7 +120,13 @@ func (e *Endpoint) updateNetworkPolicy(owner Owner, proxyWaitGroup *completion.W
 	// Publish the updated policy to L7 proxies.
-	return owner.UpdateNetworkPolicy(e, e.DesiredL4Policy, *e.prevIdentityCache, deniedIngressIdentities, deniedEgressIdentities, proxyWaitGroup)
+	var desiredL4Policy *policy.L4Policy
+	if e.desiredPolicy == nil {
```
Don't need for this, the updatenetworkpolicy is expecting `policy.L4Policy` to be a nil or not.

ianvernon replied:
But if `e.desiredPolicy` is nil, then I'd be dereferencing a nil pointer to access the `L4Policy` field, so this is necessary.

jrajahalme reviewed (approved):
Looking good, assuming you follow up on Joe's comment about hiding the copy semantics of the `EndpointPolicy` type (from desired to realized).

Commit:
endpoint: remove unnecessary setting of pointers

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
policy: hide realization of desired state in function

Signed-off by: Ian Vernon <ian@cilium.io>

aanm reviewed (commented):
@ianvernon got it, I think we should add the realized/desired policy to the API in that case.

aanm commented on pkg/endpoint/endpoint.go:
```diff
@@ -707,27 +680,48 @@ func (e *Endpoint) GetPolicyModel() *models.EndpointPolicyStatus {
 		AllowedEgressIdentities:  realizedEgressIdentities,
-		CidrPolicy:               e.L3Policy.GetModel(),
-		L4:                       e.RealizedL4Policy.GetModel(),
+		CidrPolicy:               realizedCIDRPolicy.GetModel(),
```
Oh never mind, I thought this was the desired and the line right after was the realized policy.
